pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller.go

```diff
@@ -0,0 +1,227 @@
+		masterNodeLister:           masterNodeLister,
+		masterMachineLister:        masterMachineLister,
+		networkLister:              networkLister,
+		hasNetworkInformerSyncedFn: hasNetworkInformerSyncedFn,
+	}
+	return factory.New().
+		WithSync(c.sync).
+		ResyncEvery(33*time.Minute). // make it slow since nodes are updated every few minutes
+		WithInformers(
+			masterNodeInformer,
+			masterMachineInformer,
+		).ToController("ClusterMemberRemovalController", eventRecorder.WithComponentSuffix("cluster-member-removal-controller"))
+}
+
+func (c *ClusterMemberRemovalController) sync(ctx context.Context, _ factory.SyncContext) error {
+	if !c.hasNetworkInformerSyncedFn() {
+		return nil
+	}
+	// stop if the machine API is not functional
+	if isFunctional, err := c.isMachineAPIFunctionalFn(); err != nil {
+		return err
+	} else if !isFunctional {
+		return nil
+	}
+	// we are only interested in unhealthy members
+	// because theirs nodes are gone
+	unhealthyMembers, err := c.etcdClient.UnhealthyMembers(ctx)
+	if err != nil {
+		return err
+	}
+	if len(unhealthyMembers) == 0 {
+		return nil
+	}
+	for _, unhealthyMember := range unhealthyMembers {
+		memberURL, err := memberToURL(unhealthyMember)
+		if err != nil {
+			klog.Errorf("%v", err)
+			continue
+		}
+		memberLocator := fmt.Sprintf("%v/%v/%v", memberURL, unhealthyMember.Name, unhealthyMember.ID)
+
+		machine, err := c.getMachineForMember(unhealthyMember)
+		if err != nil {
+			klog.Errorf("unable to get a machine for the unhealthy member: %v, err: %v", memberLocator, err)
+			continue
+		}
+		if machine != nil {
+			klog.V(4).Infof("cannot remove the unhealthy member: %v because a machine resource still exists (%v/%v)", memberLocator, machine.Name, machine.UID)
+			continue
+		}
+
+		node, err := c.getNodeForMember(unhealthyMember)
+		if err != nil {
+			klog.Errorf("unable to get a node for the unhealthy member: %v, err: %v", memberLocator, err)
+			continue
+		}
+		if node != nil {
+			klog.V(4).Infof("cannot remove the unhealthy member: %v because a node resource still exists (%v/%v)", memberLocator, node.Name, node.UID)
+			continue
+		}
+
+		if len(unhealthyMember.Name) == 0 {
+			klog.Errorf("unable to remove the unhealthy member: %v because it doesn't have a name (has the member been started?)", memberLocator)
+			continue
+		}
+		err = c.etcdClient.MemberRemove(ctx, unhealthyMember.Name)
+		if err != nil {
+			klog.Errorf("failed to remove the unhealthy member: %v, err: %v", unhealthyMember.Name, err)
+		} else {
+			klog.V(2).Infof("successfully removed the unhealthy member: %v from the cluster", memberLocator)
+		}
+	}
+	return nil
+}
+
+func (c *ClusterMemberRemovalController) getMachineForMember(member *etcdserverpb.Member) (*machinev1beta1.Machine, error) {
+	node, err := c.getNodeForMember(member)
+	if err != nil {
+		return nil, err
+	}
+
+	// node is gone, make sure we don't have a machine with
+	// the same internal IP as the node would have had
+	if node == nil {
```

Comment:
@JoelSpeed is this a safe assumption? That is being able to find a machine by an internal IP.

Reply:
Yes it should be, we sync the addresses in the Machine with those that the cloud provider tells us, when a Kubelet requests a serving cert it includes all of the Node addresses, we check that the Machine and Node lists match when issuing the cert so this should be safe based on that.

When the host disappears from the cloud side, we stop updating the Machine so the addresses and information should persist there as a tombstone as the Machine is marked Failed